### ISSUE TITLE
Add the iCub 3 model

### ISSUE THREAD

**Issue by S-Dafarra** [robotology/icub-models#41]
In order to start performing experiments with the iCub 3 robot, it is fundamental to have its urdf and the possibility of using it in Gazebo.

With this issue, I would like to track the generation and the issues arising when trying to use this model for the first time.

cc @pattacini, @traversaro, @fiorisi

**Comment by pattacini** [comment]
@S-Dafarra a quick update on this.

Our responsible person for generating the URDF is @salvi-mattia 👨🏻‍💼
We've organized an internal course to transfer the knowledge of exporting models that will take place next week.

More in general, we're set to proceed iteratively aiming at the following MVPs:
1. URDF of Torso + Head.
1. URDF of Torso + Head + Limbs.
1. URDF of The whole robot including fingers.

The exact number, as well as the content of MVPs, will be suitably detailed/discussed throughout the process.

Also, we will invite you periodically to attend our internal updates.

cc @DanielePucci @maggia80

**Comment by S-Dafarra** [comment]
Sounds good, thanks @pattacini

**Comment by pattacini** [comment]
Hi @S-Dafarra 

We've just invited you to `icub-tech-iit/cad-mechanics` where @salvi-mattia will be opening up a series of issues to address activities on the CAD side. You can follow the progress there.

We'll be using this issue instead to report on specific news regarding the URDF _per se_.

**Comment by lornat75** [comment]
cc @xEnVrE

**Comment by salvi-mattia** [comment]
hi everyone,
[these](https://github.com/icub-tech-iit/cad-mechanics/wiki/Creo-Mechanism-to-URDF) are the guidelines we will follow building up the URDF files foriCub3.
In the next weeks I'll work on the cad related part.
Following the handson and talking with @fiorisi we think in could be useful to generate different URDF related tu anatomical robot'a parts and assemble them at URDF level.
[ie different icubs share the same head, we often build partial robots, it could be possible to change the hands as they will be developed]
So we thought about this modules:
- torso, shoulders and upperarms
- head
- legs
- hands and forearms
we will need to define "universally accepted" interfaces between the modules

@S-Dafarra @FabioBergonti @fiorisi any opinions?

**Comment by fiorisi** [comment]
I agree @salvi-mattia.
If there is a simple way ([xacro](http://wiki.ros.org/xacro)?) to reassemble the URDF modules is the best way to handle different robot versions and upgrades (e.g. new foreamr-hand system).

**Comment by traversaro** [comment]
I think it is a great idea! I am not sure about what is the best way of composing the part URDF to generate a complete URDF, but we can sort that out.

**Comment by salvi-mattia** [comment]
[Here](https://github.com/icub-tech-iit/cad-mechanics-public/wiki/Creo-mechanism-to-URDF) you can find the last version of the procedure.

It is public and it will be the official version for this procedure, we will keep this one up to date

**Comment by DanielePucci** [comment]
CC @FabioBergonti @vpunithreddy @gabrielenava

**Comment by salvi-mattia** [comment]
looking @ iCub2 simulation model, the head just include the three DOF in the neck, do we wanto to have also eyes' movementes?

**Comment by DanielePucci** [comment]
Hi @salvi-mattia, my take is that in the long run for sure, but I would say that they are not necessary for the first simulations of whole-body dynamic control and locomotion. So, if the process of adding the eyes' movements slows down considerably (e.g. days) the release date of the model, I would say to avoid adding them unless the DOFs are necessary in the short term to other research lines

CC
@lornat75 @ChiaraBartolozzi

**Comment by pattacini** [comment]
The eyes are traditionally a dull pain to deal with in Gazebo (@xEnVrE knows it quite well), so we definitely need to address the automatic generation.

That being said, I agree that for the first MVP we could skip them and proceed faster to then refine over if those small mechanical details regarding eyes don't turn out to be a walk in the park.

**Comment by salvi-mattia** [comment]
designing robots is traditionally way better than a walk in the park.

**Comment by salvi-mattia** [comment]
I have fixed some fail in the iCub3 model,
at the moment it weighs 55 kg [in the CAD]
while the actual robot is 51 kilos
this error is not that easy to fix because it's not easy to spot where the problem is really located.

do we think we could go ahead with this mass properties?

@S-Dafarra @fiorisi @DanielePucci

**Comment by pattacini** [comment]
Nicely done @salvi-mattia 👍🏻 
I remember those old times when the weight discrepancy was used to amount to tens of kilos 😄 

My hunch is that we could proceed as it is now and then once we'll have the first overall exportable model we can backtrace and bisect the CAD in order to narrow down the problem. This means that we will very likely need eventually to "bisect" the real robot in order to measure the weights of the single parts as well.

**Comment by fiorisi** [comment]
@salvi-mattia for the iCub2.5 model we had to ["manually" adjust the weight](https://github.com/robotology/icub-model-generator/blob/9f6dbe7ab98c4a4762317e2c9aacb298ffe39c53/simmechanics/data/icub2_5/ICUB_2-5_BB_simmechanics_options.yaml.in#L932-L939) of the robot by reducing the mass of the sub assemblies in the second part of the URDF generation process (from the SimMechanics model to the URDF). 

See [this Excel table](https://github.com/robotology/icub-model-generator/blob/master/simmechanics/data/icub2_5/ConversionProcess.xlsx) to have an idea.

For this it would be nice to have an idea about the weight error for the main sub-assemblies (e.g. lower body, upper body and head).

[In this issue](https://github.com/robotology/icub-tech-support/issues/98), some of the robot sub-assemblies were weighted. It should be already possible to check the error.

**Comment by fiorisi** [comment]
One of the [prerequisites](https://github.com/icub-tech-iit/cad-mechanics-public/wiki/Creo-mechanism-to-URDF#preliminary-operations) is to have a clear list of joint and link names. Something like [this](http://wiki.icub.org/wiki/ICub_Model_naming_conventions) (for iCub2). 

Otherwise, later on we will have to rename many files and change the associated script files.

**Comment by salvi-mattia** [comment]
@S-Dafarra if you do have a list like that to provide it could be useful.
otherwise I'm naming the link following icub2 style and we will compose the table at the end [if smthg will be terribly wrong we con always rename some stuff]

**Comment by DanielePucci** [comment]
@salvi-mattia 
> at the moment it weighs 55 kg [in the CAD]
while the actual robot is 51 kilos
this error is not that easy to fix because it's not easy to spot where the problem is really located.
>
> do we think we could go ahead with this mass properties?

I would love to weigh 4 kilos less than what my CAD model say 😄 With that said, I agree with @fiorisi  on

> [In this issue](https://github.com/robotology/icub-tech-support/issues/98), some of the robot sub-assemblies were weighted. It should be already possible to check the error.

So, if this process does not take too long, we would love to have already an idea where the main weight mismatch (CAD vs Reality) is. Consider that we could always unmount the robot (e.g. split it in two) and weigh it if we are in trouble (CC @Fabrizio69 @maggia80).

Unfortunately, back in time, *accurate weight* in the CAD was one of the important features to have in order to  achieve the [first balancing results](https://www.youtube.com/watch?v=jaTEbCsFp_M), and the mismatch that @pattacini [mentioned](https://github.com/robotology/icub-models/issues/41#issuecomment-674090060) was among the things that we had to correct. 

So, having an idea on where to correct the weight would later help the control algorithms, which make use of the CAD information extensively for (among other things) [center of pressure](https://en.wikipedia.org/wiki/Center_of_pressure_(terrestrial_locomotion)) and [center of mass](https://en.wikipedia.org/wiki/Center_of_mass) computations.

**Comment by salvi-mattia** [comment]
hi all,
iCub3 model has a plenty of simp rep whose reliability we arn not sure.
I started creating a new bunch called Sim_Icub3_[link name]
es. Sim_Icub3_Chest
Sim_Icub3_R Elbow
Sim_Icub3_Shoulder_1
Sim_Icub3_Shoulder_2
Sim_Icub3_Upperarm

they still miss datums [axis, frames or planes] for further assembly

@FabioBergonti @fiorisi @Mick3Lozzo 
if you could please give a fast look at them we can check if we all understand the same concepts in creating sim reps

It could be time saving to spot systematic errors in the early part of the work.

@S-Dafarra i dont know if you have acces to the cad or someone can show you what we are talking about.

**Comment by pattacini** [comment]
Yep, @S-Dafarra can read the CAD drawings.

**Comment by salvi-mattia** [comment]
kudos 4 @FabioBergonti that make me notice that I didn't include in the "chest" part the small part of the neck that "doesn't move" but it's fixed with the sholuder.
that is true and that is because the connection between the chest and the head is not in a joint but "in the middle" of a link
to fix thi minor problem I suggest to use a "fixed" joint between these two parts [as it happend for the FTsensor that is a "fixed joint" in the middle of a link]

do you think it could be a good way to proceed?

@fiorisi @S-Dafarra

**Comment by S-Dafarra** [comment]
> @S-Dafarra if you do have a list like that to provide it could be useful.
> otherwise I'm naming the link following icub2 style and we will compose the table at the end [if smthg will be terribly wrong we con always rename some stuff]

I don't have a list at the moment. When writing code, we tried to be as generic as possible. Clearly, having a list of joints similar to iCub2 would simplify the porting process. One thing that we usually assume is that there is a frame on the robot sole having the z axis pointing upward and the x axis pointing forward. These frames are called ``l_sole`` and ``r_sole`` and are kind of omnipresent in our codes. Another thing pretty common is the ``root_link``. Apart from these few features, we are pretty open for the naming.

> @S-Dafarra I don't know if you have access to the cad or someone can show you what we are talking about.

I do have access to the repo, but I am not sure if I can provide any useful comment on this :grin: For sure I appreciate any effort in making the process of generating URDF models more reliably. 

> do you think it could be a good way to proceed?

I don't think it is a problem having an additional fixed joint. In view of having the URDF composable, my two cents would be to include the blue link to the head subassembly if this part can be changed according to the installed head, to the chest otherwise.

**Comment by salvi-mattia** [comment]
I think Ihave integrated different modifications suggested by @FabioBergonti 

as we said [here](https://github.com/robotology/icub-models/issues/41#issuecomment-670546542)

I think we have now the simp reps to complete the first module: **torso, shoulders and upperarms**

@fiorisi if you could please give a fast check we will procede adding datums and making shrinkwrap
so we can start validating the process in the next [non-CAD] steps



![immagine](https://user-images.githubusercontent.com/43994292/90409426-6a8de380-e0a9-11ea-9f69-78c82a27ac36.png)

**Comment by fiorisi** [comment]
Great @salvi-mattia! 🚀 
I had a quick look at the simplified rep and to me they look fine. Just a few comments:
- `iCub3 model has plenty of simp rep whose reliability we are not sure about.` I would delete all the unnecessary `Sim_` SimpRep.
- `I started creating a new bunch called Sim_Icub3_[link name]`. For the new SimpReps let's follow the [naming convention](https://github.com/icub-tech-iit/cad-mechanics-public/wiki/Creo-mechanism-to-URDF#how-to-define-the-simplified-representations).
- `I suggest to use a "fixed" joint between these two parts` the lower neck is part of the chest, I wouldn't divide it from the neck.

**Comment by fiorisi** [comment]
> I suggest to use a "fixed" joint between these two parts the lower neck is part of the chest, I wouldn't divide it from the neck.

After a T2T meeting with @salvi-mattia we agreed that this division is the interface between the head assembly URDF and upperbody assembly URDF.

**Comment by salvi-mattia** [comment]
together with @fiorisi we are facing various problems in naming conventions, simp rep separations, joint naming...
we are trying to overcome them not only for the contingency but to create a more reliable procedure [hoping it will help to be faster in the future]

@fiorisi @S-Dafarra fix is glad to invite you to our review in order to take stock of the situation.
it will take few minutes [15 I guess] starting from 2pm next friday [28th of august 2020]

**Comment by S-Dafarra** [comment]
> @fiorisi @S-Dafarra fix is glad to invite you to our review in order to take stock of the situation.
> it will take few minutes [15 I guess] starting from 2pm next friday [28th of august 2020]

Thanks for the invitation, but unfortunately I will be busy at that time.

**Comment by pattacini** [comment]
@S-Dafarra our review meeting will start at 2:00 pm and is supposed to end at 4:30 pm, so you happen to get a free 15 mins slot in this range, feel free to get back to us.

**Comment by fiorisi** [comment]
> @fiorisi @S-Dafarra fix is glad to invite you to our review in order to take stock of the situation.
it will take few minutes [15 I guess] starting from 2pm next friday [28th of august 2020]

Thanks @salvi-mattia, `dev` has also the review meeting at the same time.

**Comment by lornat75** [comment]
good job! I agree that having the eyes in the model is not urgent for the balancing tasks, however, it would be nice to have them.

**Comment by S-Dafarra** [comment]
> @S-Dafarra our review meeting will start at 2:00 pm and is supposed to end at 4:30 pm, so if you happen to have a free 15 mins slot in this range, feel free to get back to us.

Actually, I realized I confused this Friday with the next one. I will definitely pass by today.

**Comment by pattacini** [comment]
So we're gonna start off w/ iCub3 URDF at 14:00 if it's ok.
cc @S-Dafarra @salvi-mattia 

@xEnVrE if you want to attend too you're more than welcome, although we're not going to cover details like fingers and eye bulbs at this stage that you're more interested in.

**Comment by xEnVrE** [comment]
@pattacini thanks for the invitation. I have two meetings in that slot but if the second ends earlier I will join for sure!

**Comment by salvi-mattia** [comment]
We have completed the mechanism for the first assembly.
as we said [here](https://github.com/robotology/icub-models/issues/41#issuecomment-670546542) it represents torso, shoulders and upperarms.

@Nicogene is going to perform folloeing steps in the instructions pipeline and giving us feedback about his work and the problems he will face.

@fiorisi and @FabioBergonti have given a big help in naming links and joint and doublechecked my work step by step

@fiorisi is glad to help @Nicogene and whoevere else will work on this activity for the mechanical side.

if you update cad mechanics repository, you can find the mechanism assembly at this location

cad-mechanics\projects\simulation_model\icub3

opening it you can finally push the magic matlab button

**Comment by Nicogene** [comment]
# First iteration of urdf generation

Today I opened the asm created by @salvi-mattia for the upperbody and I pressed the magic button.
Despite a warning of Creo I obtained a list of stl and an xml file that I used for obtaining a first version of the urdf.

I simply run:

```
simmechanics_to_urdf SIM_ICUB3_UPPERBODY.xml --yaml icub3-upperbody-options.yaml --outputfile test.urdf
```

Where in the yaml I simply specified the scale factor of 0.001(Creo uses mm, gazebo m)

And here is the result, the upperbody sunbathing in gazebo 🏖️ 

![immagine](https://user-images.githubusercontent.com/19152494/91724108-1997e800-eb9d-11ea-8735-7cfc0f404ee3.png)

The next steps will include for sure the definition of joint limits through the csv file and maybe the renaming the link?

**Comment by Nicogene** [comment]
FYI the work is carried on this branch https://github.com/robotology/icub-model-generator/tree/feat/addiCub3Upperbody

**Comment by salvi-mattia** [comment]
As we decided today during fix's sprint planning I'm going to produce simp rep, shrinkwrap and mechanism for the legs, later on we will check if it's better to assemble them to the torso at CAD or URDF level.

In the mean while we will check how the ftsensor in the upperarms and its fixed joints works in this procedure

**Comment by Nicogene** [comment]
After a meeting with @salvi-mattia @FabioBergonti and @traversaro, it emerged the necessity of having the link reference frames in the CAD before the conversion to the urdf.
If not defined it is generated with the z-axis passing through the joint axis, with y-x direction arbitrary.
Moreover, different generations can give different axis configurations.

For these reasons before proceeding with the work on the legs, we would first define those frames [following this convention](https://www.ros.org/reps/rep-0103.html#axis-orientation).
With this naming convention: `SCSYS_<child_link_name>`.

Is it ok for you?

cc @pattacini @DanielePucci @S-Dafarra @maggia80

**Comment by traversaro** [comment]
Note that in the future we may need to also export a frame aligned with the traditional iCub http://wiki.icub.org/wiki/ICubForwardKinematics , but that is something that we can easily do later just for the root, while for all the other links I think it it is beneficial to use the REP 103 convention, as we already did when we clearly defined the frames for the legs in https://github.com/robotology/icub-model-generator/commit/97827f894a2572b10697a5a33639b530690151b0 .

**Comment by traversaro** [comment]
> For these reasons before proceeding with the work on the legs, we would first define those frames [following this convention](https://www.ros.org/reps/rep-0103.html#axis-orientation).
> With this naming convention: `SCSYS_<child_link_name>`.
> 
> Is it ok for you?

Note that this should avoid all the problems that we had in https://github.com/robotology/icub-models/issues/20 .

**Comment by salvi-mattia** [comment]
this is icub3 upper body seen from top
![immagine](https://user-images.githubusercontent.com/43994292/93576122-c8f4fd00-f99a-11ea-86cc-a7efe410b0bf.png)
if I correctly interpret 
![immagine](https://user-images.githubusercontent.com/43994292/93576179-e1fdae00-f99a-11ea-9353-c058517e6b1f.png)
X axis will be placed like the yellow one and not like the blue

it means that we will have a lot of joint axis coincident with no frame axis.

is it correct?
is it the way we want to proced?

@S-Dafarra @pattacini @DanielePucci @traversaro @fiorisi @FabioBergonti

**Comment by pattacini** [comment]
@salvi-mattia it sounds a bit weird.

**Comment by traversaro** [comment]
Both options have their advantages, some days ago with @salvi-mattia I defended the "yellow arrow"-option but if the plan in the long term is to consider each limb as a standalone robot, probably the option that makes more sense even in accordance to rep-103 is the "blue arrow", sorry if I changed my mind.

**Comment by S-Dafarra** [comment]
To be honest, I am not able to find particular advantages in having the yellow solution, so I would prefer the blue one and have at least one axis aligned with the axis of rotation.

Am I missing something?

**Comment by traversaro** [comment]
In general, having parallel frames in at least one pose of the robot drastically simplifies the debugging of vectors expressed in different link frame, as it is not necessary to rotate them, or for position vectors you just need to sum them to the origin of the frame, and that is the reason why most URDF assign the link orientations to be parallel to each other, as described in REP-103 .

**Comment by S-Dafarra** [comment]
> In general, having parallel frames in at least one pose of the robot drastically simplifies the debugging of vectors expressed in different link frame, as it is not necessary to rotate them, or for position vectors you just need to sum them to the origin of the frame, and that is the reason why most URDF assign the link orientations to be parallel to each other, as described in REP-103 .

Ah I see, thanks. Anyways, if the robot moves from the zero position, anyway those frames will not be aligned anymore right, especially on the yellow case?

**Comment by salvi-mattia** [comment]
talking with @traversaro it seems to be a good solution to have a definition with no uncertainty dependig on robot position.

our proposal [shared with @FabioBergonti ] could be:

Z axis on the joint axis
X axis on the link main axis [pointing next joint]
Y axis according to dextrous convention

If no one spots problem in this kind of convention we can proceed and see how it will work

**Comment by traversaro** [comment]
Just to clarify instead the advantage of the "all links in the same limb" strategy:
* It is coherent we the strategy we already adopted with the

**Comment by traversaro** [comment]
> I can list a few as soon if I have time

Humanoid models that follow the convention "all frame have the same orientation in a given configuration): 
* Boston Dynamics' Atlas: https://github.com/osrf/drcsim/blob/master/atlas_description/urdf/atlas.urdf
* NASA' Valkyrie: https://github.com/gkjohnson/nasa-urdf-robots/blob/master/val_description/model/robots/valkyrie_A.urdf
* SoftBank Robotics Nao: https://github.com/ros-naoqi/nao_robot/blob/master/nao_description/urdf/naoV50_generated_urdf/nao.urdf 
* PAL Robotics' REEM C and Talos: https://github.com/pal-robotics/talos_robot/blob/kinetic-devel/talos_description/urdf/arm/arm.urdf.xacro 
* ROBOTIS' OP3 : https://github.com/ROBOTIS-GIT/ROBOTIS-OP3-Common/blob/0780e54c2064a42089b0cf970d716ae44fec9233/op3_description/urdf/robotis_op3.structure.lleg.xacro
* Shadow Hand (not a humanoid but indeed a complex articulated model): https://github.com/shadow-robot/sr_common/blob/melodic-devel/sr_description/hand/xacro/finger/distal/distal.urdf.xacro 

The only humanoid URDF that is not using this convention is the iCub 2.5, that is using this convention for the legs but for the arms has arbitrary frame coming out from the Creo generation process (see https://github.com/robotology/icub-models/issues/20).

> >  In general, having parallel frames in at least one pose of the robot drastically simplifies the debugging of vectors expressed in different link frame, as it is not necessary to rotate them, or for position vectors you just need to sum them to the origin of the frame, and that is the reason why most URDF assign the link orientations to be parallel to each other, as described in REP-103 .
> 
> Ah I see, thanks. Anyways, if the robot moves from the zero position, those frames will not be aligned anymore right, especially on the yellow case?

Exactly.

**Comment by pattacini** [comment]
Are we maybe discussing about 3 different conventions?
- the "yellow-arrow";
- the "iCub 2.5 legs";
- the one proposed in https://github.com/robotology/icub-models/issues/41#issuecomment-694780076.

That one @traversaro seems to support is not the so-called "yellow-arrow" but rather that one already implemented in the iCub 2.5 legs. On the other hand, @salvi-mattia, yours is more "traditional".

Did I get this correct?

**Comment by traversaro** [comment]
> Did I get this correct?

Exactly. The main difference between "yellow-arrow" and "iCub 2.5 legs" is at which level you try to have link frame all with the same orientation. For  "yellow-arrow"  it is at the robot level (that given iCub shape leads to peculiar behavior) while "iCub 2.5 legs" is just to keep the same orientation for all the links in a given limb. 

>  X axis on the link main axis [pointing next joint]

Small note: with this convention you still have have a small ambiguity (the same that you have in DH) when assigning frames for a joint whose axis coincide (such as the ankle or hip in iCub).

**Comment by kouroshD** [comment]
> Humanoid models that follow the convention "all frame have the same orientation in a given configuration):

I agree with @traversaro . Back in time, when I was using the different robot's urdf for the retargeting purposes, I had this problem with iCub2.5 urdf model. Instead, other robots were straight forward to do retargeting.

I do not agree with this comment https://github.com/robotology/icub-models/issues/41#issuecomment-694780076 , since the robot motor axis can be different, and we will end up with urdf model that every link has different frame orientations.  But instead, putting all the joint to N-pose or T-pose (or zero values for all the motor joint angles) we will have a much cleaner urdf of the robot.

**Comment by traversaro** [comment]
Hi @S-Dafarra @kouroshD , if you could discuss internally with other DIC people and reach a consensus on this it would be great, otherwise the risk is that the discussion become a bit too dispersive.

**Comment by S-Dafarra** [comment]
> Hi @S-Dafarra @kouroshD , if you could discuss internally with other DIC people and reach a consensus on this it would be great, otherwise the risk is that the discussion become a bit too dispersive.

Sorry to show up just now. We discussed internally and we agreed on the following. The convention we prefer is defined as follows:
- When all the joints are in the zero position, all the links belonging to a limb share the same orientation. E.g., all the links of the torso would be aligned with frame A, all the links of the left arm with frame B, the right arm with frame C,....
- These "common" frames (A, B, C, ...) have the z-axis pointing upward, the x-axis pointing forward (clearly, upward and forward depend on the limb)

Given the particular design of the shoulders, in the zero position, a frame on an arm link is slightly rotated around the z-axis compared to a frame in the torso.

This convention is basically the one defined as "iCub 2.5 legs" in https://github.com/robotology/icub-models/issues/41#issuecomment-694884201, which corresponds to the blue arrow of https://github.com/robotology/icub-models/issues/41#issuecomment-694738366.

**Comment by salvi-mattia** [comment]
after a meeting with @S-Dafarra @FabioBergonti and @pattacini 
we agreed on the following:
- every link has a solidal frame with 
- X axis pointing forward
- Z axis pointing up
- Y axis to have a dextrous frame
- one of the axis has to be coincident with the joint axis

when it's not applicable (in plaine words, we may derive a better formalism later):
- one axis is coincident with joint axis
- Z axis is the one pointing upwards at the most
- X axis pointing forward at the most
- Y axis to have a dextrous frame

**Comment by salvi-mattia** [comment]
![immagine](https://user-images.githubusercontent.com/43994292/94129635-e541df80-fe5b-11ea-9127-e0cdd6ad2ea8.png)

**Comment by salvi-mattia** [comment]
I added the frames in the linke and reassembled the mechanism.
Everything is pushed in cad-mechanics
@Nicogene u can now make your magic-button-stuff

after that it could be useful to check if we are getting what we want, so that we can proceed with the legs in the correct way

**Comment by Nicogene** [comment]
@S-Dafarra in [icub3-upperbody.zip](https://github.com/robotology/icub-models/files/5311133/icub3-upperbody.zip) you should find all you need for opening the urdf in gazebo, can you check if the frames are following the right convention?

**Comment by S-Dafarra** [comment]
> @salvi-mattia added and checked the link frames but 4 eyes are better than two 
> 
> @S-Dafarra could you please double-check if the frames are following the right convention?
> 
> In [icub3-upperbody.zip](https://github.com/robotology/icub-models/files/5311133/icub3-upperbody.zip) you should find all you need for opening the urdf in gazebo.
> Please let me know if you experience any issue.

Hi @Nicogene, I just tried, but I am not able to load the model. I get the following error
```
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_root_link_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_torso_1_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_torso_2_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_chest_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_l_shoulder_1_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_l_shoulder_2_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_l_shoulder_3_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_l_upperarm_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_r_shoulder_1_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_r_shoulder_2_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_r_shoulder_3_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [sim_icub3_r_upperarm_prt.stl]
[Err] [Visual.cc:2951] No mesh specified
```
I just specified the ``GAZEBO_MODEL_PATH``, should I add something else?

**Comment by Nicogene** [comment]
For simplicity run `gazebo` in the folder where the stl files are located.

This issue should not occur if the files are installed via CMake in the right location.

**Comment by S-Dafarra** [comment]
> For simplicity run `gazebo` in the folder where the stl files are located.
> 
> This issue should not occur once CMake will install them in the right location.

Thanks, that worked



> @S-Dafarra could you please double-check if the frames are following the right convention?

Actually, there seems to be a problem. It seems that the shoulders motor axis are not aligned to any axis

![image](https://user-images.githubusercontent.com/18591940/94798813-70821e80-03e2-11eb-8c12-93a58d3f2b96.png)

It seems it is not following what was defined in https://github.com/robotology/icub-models/issues/41#issuecomment-697561933 but rather the "yellow arrow" convention of https://github.com/robotology/icub-models/issues/41#issuecomment-694738366

**Comment by Nicogene** [comment]
Sorry probably I used the wrong yaml in the generation 🙏 
Stay tuned !

**Comment by salvi-mattia** [comment]
oh, well, my fault too
I named all the frame `CSYS_<child_link_name>`. instead of `SCSYS_<child_link_name>`.

lets have a check with ths wrong naming convention and then I'll fix it [if we think it's worth it]

**Comment by Nicogene** [comment]
Now seems ok to me

Please @S-Dafarra could you check it now?

[icub3-upperbody-final.zip](https://github.com/robotology/icub-models/files/5312178/icub3-upperbody-final.zip)

**Comment by S-Dafarra** [comment]
> Now seems ok to me
> 
> Please @S-Dafarra could you check it now?
> 
> [icub3-upperbody-final.zip](https://github.com/robotology/icub-models/files/5312178/icub3-upperbody-final.zip)
> 

Looks good to me, thanks a lot

![image](https://user-images.githubusercontent.com/18591940/94814706-f78ec100-03f9-11eb-9a17-ee0638f37691.png)

cc @kouroshD

**Comment by salvi-mattia** [comment]
Plot twist,
the feet mounted on iCub3 are the same that in the cad: onf single FTsensor and passvie fingers with spring.
I'm aware that exist a new project with different feet [not mounted]
Which is the state of this enhancement? @fiorisi @Fabrizio69 @Mick3Lozzo 

which kind of feet do we want in the URDF? @S-Dafarra

**Comment by S-Dafarra** [comment]
> which kind of feet do we want in the URDF? @S-Dafarra

Preferably the one of the new project, which should be rectangular without the passive toe, holding two FTs.

**Comment by Fabrizio69** [comment]
>
>Which is the state of this enhancement? @fiorisi @Fabrizio69 @Mick3Lozzo
>

The state is that the new feet are ready to be mounted on the robot (see https://github.com/icub-tech-iit/tickets/issues/224#issuecomment-645229863). There are three possible sizes for each foot plate:
100 mm
110 mm
120 mm
The simplified feet replacing kit is cod.14658 - IG_022_A_001. At the moment the kit is not mounted on the iCub3 CAD.

**Comment by Nicogene** [comment]
The https://github.com/robotology/icub-model-generator/pull/133 has been merged.

The model should be available 

This is a first draft of the model, it has some open points/issues that we worked-around but I think it is something usable.

Here it is the `iCub 3 - Venere di Milo edition`

![immagine](https://user-images.githubusercontent.com/19152494/96244511-7103e300-0fa6-11eb-9807-e2d335b96146.png)

We added also the frames `l_sole` and `r_sole` as requested.
<details>
<summary>🔘 Click here to see the output of `idyntree-model-info` </summary>

```
Model:
  Links:
    [0] root_link
    [1] l_hip_1
    [2] r_hip_1
    [3] torso_1
    [4] torso_2
    [5] chest
    [6] l_shoulder_1
    [7] r_shoulder_1
    [8] neck_1
    [9] neck_2
    [10] neck_3
    [11] head
    [12] r_shoulder_2
    [13] r_shoulder_3
    [14] r_upper_arm
    [15] l_shoulder_2
    [16] l_shoulder_3
    [17] l_upper_arm
    [18] r_hip_2
    [19] r_hip_3
    [20] r_upper_leg
    [21] r_lower_leg
    [22] r_ankle_1
    [23] r_ankle_2
    [24] r_foot_front
    [25] r_foot_rear
    [26] l_hip_2
    [27] l_hip_3
    [28] l_upper_leg
    [29] l_lower_leg
    [30] l_ankle_1
    [31] l_ankle_2
    [32] l_foot_rear
    [33] l_foot_front
  Frames:
    [34] base_link --> root_link
    [35] r_sole --> r_foot_rear
    [36] l_sole --> l_foot_rear
    [37] l_arm_ft_sensor --> l_upper_arm
    [38] r_arm_ft_sensor --> r_upper_arm
    [39] l_leg_ft_sensor --> l_hip_3
    [40] l_foot_front_ft_sensor --> l_foot_front
    [41] l_foot_rear_ft_sensor --> l_foot_rear
    [42] r_leg_ft_sensor --> r_hip_3
    [43] r_foot_front_ft_sensor --> r_foot_front
    [44] r_foot_rear_ft_sensor --> r_foot_rear
  Joints:
    [0] l_hip_pitch (dofs: 1) : root_link<-->l_hip_1
    [1] r_hip_pitch (dofs: 1) : root_link<-->r_hip_1
    [2] torso_roll (dofs: 1) : root_link<-->torso_1
    [3] torso_pitch (dofs: 1) : torso_1<-->torso_2
    [4] torso_yaw (dofs: 1) : torso_2<-->chest
    [5] l_shoulder_pitch (dofs: 1) : chest<-->l_shoulder_1
    [6] r_shoulder_pitch (dofs: 1) : chest<-->r_shoulder_1
    [7] neck_pitch (dofs: 1) : neck_1<-->neck_2
    [8] neck_roll (dofs: 1) : neck_2<-->neck_3
    [9] neck_yaw (dofs: 1) : neck_3<-->head
    [10] r_shoulder_roll (dofs: 1) : r_shoulder_1<-->r_shoulder_2
    [11] r_arm_ft_sensor (dofs: 1) : r_shoulder_3<-->r_upper_arm
    [12] l_shoulder_roll (dofs: 1) : l_shoulder_1<-->l_shoulder_2
    [13] l_arm_ft_sensor (dofs: 1) : l_shoulder_3<-->l_upper_arm
    [14] r_hip_roll (dofs: 1) : r_hip_1<-->r_hip_2
    [15] r_hip_yaw (dofs: 1) : r_hip_3<-->r_upper_leg
    [16] r_knee (dofs: 1) : r_upper_leg<-->r_lower_leg
    [17] r_ankle_pitch (dofs: 1) : r_lower_leg<-->r_ankle_1
    [18] r_ankle_roll (dofs: 1) : r_ankle_1<-->r_ankle_2
    [19] l_hip_roll (dofs: 1) : l_hip_1<-->l_hip_2
    [20] l_hip_yaw (dofs: 1) : l_hip_3<-->l_upper_leg
    [21] l_knee (dofs: 1) : l_upper_leg<-->l_lower_leg
    [22] l_ankle_pitch (dofs: 1) : l_lower_leg<-->l_ankle_1
    [23] l_ankle_roll (dofs: 1) : l_ankle_1<-->l_ankle_2
    [24] neck_fixed_joint (dofs: 0) : chest<-->neck_1
    [25] r_shoulder_yaw (dofs: 0) : r_shoulder_2<-->r_shoulder_3
    [26] l_shoulder_yaw (dofs: 0) : l_shoulder_2<-->l_shoulder_3
    [27] r_leg_ft_sensor (dofs: 0) : r_hip_2<-->r_hip_3
    [28] r_foot_front_ft_sensor (dofs: 0) : r_ankle_2<-->r_foot_front
    [29] r_foot_rear_ft_sensor (dofs: 0) : r_ankle_2<-->r_foot_rear
    [30] l_leg_ft_sensor (dofs: 0) : l_hip_2<-->l_hip_3
    [31] l_foot_rear_ft_sensor (dofs: 0) : l_ankle_2<-->l_foot_rear
    [32] l_foot_front_ft_sensor (dofs: 0) : l_ankle_2<-->l_foot_front
```
</details>


You should find it on `devel` branch.

I would keep this issue open until we don't have the complete model.

cc @S-Dafarra @DanielePucci @maggia80 @salvi-mattia @pattacini @fiorisi @traversaro @FabioBergonti

**Comment by S-Dafarra** [comment]
Thanks a lot @Nicogene @salvi-mattia and every one, this is something from which we can start!

**Comment by S-Dafarra** [comment]
I have tried to change the branch and install all robots. I can see iCubGenova09 in the list of models, but it fails to load it. These are the errors I get
```
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_root_link_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_torso_1_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_torso_2_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_chest_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_l_shoulder_1_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_l_shoulder_2_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_l_shoulder_3_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_l_upperarm_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_r_shoulder_1_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_r_shoulder_2_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_r_shoulder_3_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
[Wrn] [SystemPaths.cc:459] File or path does not exist [""] [urdf file/urdf file/sim_icub3_r_upperarm_prt.stl]
[Err] [Visual.cc:2953] No mesh specified
```

**Comment by traversaro** [comment]
> I have tried to change the branch and install all robots. I can see iCubGenova09 in the list of models, but it fails to load it.

I guess in the yaml of simmechanics-to-urdf the `filenameformatchangeext` used to specify where the meshes are installing is missing, so the mesh elements in the URDF have `<mesh filename="sim_icub3_torso_1_prt.stl" scale="0.001 0.001 0.001"/>` instead of the correct path as in the iCub 2 URDFs.

**Comment by Nicogene** [comment]
The stl problem should be fixed by https://github.com/robotology/icub-model-generator/pull/145.

(Tested it successfully)

**Comment by traversaro** [comment]
> The stl problem should be fixed by [robotology/icub-model-generator#145](https://github.com/robotology/icub-model-generator/pull/145).
> 
> (Tested it successfully)

Merged! The deployment time from icub-model-generator to icub-models can be probably be improved by migrating from Travis to GitHub Actions and using caching, but that's another story.

**Comment by traversaro** [comment]
> 
> 
> > The stl problem should be fixed by [robotology/icub-model-generator#145](https://github.com/robotology/icub-model-generator/pull/145).
> > (Tested it successfully)
> 
> Merged! The deployment time from icub-model-generator to icub-models can be probably be improved by migrating from Travis to GitHub Actions and using caching, but that's another story.

Model deployed after ~30 minutes, ~13 minutes of queue and 17 minutes of CI: https://github.com/robotology/icub-models/commit/29f5fddad7f9da4dbebc67e2fac13f9dcb2aaceb .

**Comment by S-Dafarra** [comment]
I am still not able to load the model in ``gazebo``. ``iCubV3`` loads only the upperbody.

**Comment by traversaro** [comment]
Just to check, does `iCubGenova09` works well instead? If you run `gazebo --verbose` reveals something?

**Comment by traversaro** [comment]
Note that there still present the regression https://github.com/robotology/icub-models/issues/50 that could create problems.

**Comment by S-Dafarra** [comment]
> Just to check, does `iCubGenova09` works well instead? If you run `gazebo --verbose` reveals something?

After a couple of reinstallation of the ``robotology-superbuild`` I see this
![image](https://user-images.githubusercontent.com/18591940/96334807-ba6e3400-1073-11eb-839e-4b30a5a44715.png)

The same for ``iCubV3``. Here the output of the terminal
```
sdafarra@iiticublap104:~$ gazebo -slibgazebo_yarp_clock.so --verbose
Gazebo multi-robot simulator, version 11.2.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[Msg] Waiting for master.
Gazebo multi-robot simulator, version 11.2.0
Copyright (C) 2012 Open Source Robotics Foundation.
Released under the Apache 2 License.
http://gazebosim.org

[DEBUG] GazeboYarpClock loaded. Clock port will be  /clock
[Msg] Waiting for master.
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 192.168.1.84
[INFO] |yarp.os.Port| Port /clock active at tcp://192.168.1.84:10002/
[INFO] |yarp.os.Port| Port /clock/rpc active at tcp://192.168.1.84:10003/
[Msg] Connected to gazebo master @ http://127.0.0.1:11345
[Msg] Publicized address: 192.168.1.84
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/isaac/isaac-wp1/EVAmodelFiles/generatedUrdfs/robots/isaacEvaYarp_all"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/isaac/isaac-wp1/EVAmodelFiles/generatedUrdfs"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/isaac/isaac-wp1/EVAmodelFiles/leo_description"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/gazebo/models/.git"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/iCub/robots/icubGazeboSim"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/BlockFactory"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/ICUBcontrib"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/WBToolbox"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/YCM"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/bash-completion"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/gazebo"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/iCub"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/robotology-superbuild"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/wbc"
[Err] [InsertModelWidget.cc:402] Missing model.config for model "/home/sdafarra/Software/robotology-superbuild/build/install/share/yarp"
[INFO] |yarp.devices.AnalogWrapper| No ROS group found in config file ... skipping ROS initialization.
[INFO] |yarp.os.Port| Port /icubSim/left_arm/analog:o/rpc:i active at tcp://192.168.1.84:10004/
[INFO] |yarp.devices.AnalogWrapper| AnalogServer : no ROS initialization required
[INFO] |yarp.dev.PolyDriver| Created wrapper <analogServer>. See C++ class AnalogWrapper for documentation.
[INFO] |yarp.dev.PolyDriver| Created device <gazebo_forcetorque>. See C++ class GazeboYarpForceTorqueDriver for documentation.
[INFO] |yarp.os.Port| Port /icubSim/left_arm/analog:o active at tcp://192.168.1.84:10005/
[Wrn] [msgs.cc:1842] Conversion of sensor type[force_torque] not supported.
[Wrn] [msgs.cc:1842] Conversion of sensor type[force_torque] not supported.
[INFO] |yarp.devices.AnalogWrapper| No ROS group found in config file ... skipping ROS initialization.
[INFO] |yarp.os.Port| Port /icubSim/right_arm/analog:o/rpc:i active at tcp://192.168.1.84:10006/
[INFO] |yarp.devices.AnalogWrapper| AnalogServer : no ROS initialization required
[INFO] |yarp.dev.PolyDriver| Created wrapper <analogServer>. See C++ class AnalogWrapper for documentation.
[INFO] |yarp.dev.PolyDriver| Created device <gazebo_forcetorque>. See C++ class GazeboYarpForceTorqueDriver for documentation.
[INFO] |yarp.os.Port| Port /icubSim/right_arm/analog:o active at tcp://192.168.1.84:10007/
[Wrn] [msgs.cc:1842] Conversion of sensor type[force_torque] not supported.
[Wrn] [msgs.cc:1842] Conversion of sensor type[force_torque] not supported.
[ERROR] |yarp.devices.ControlBoardWrapper| Parameter networks use deprecated syntax
[INFO] |yarp.devices.ControlBoardWrapper| /icubSim/torso : no ROS initialization required
[INFO] |yarp.devices.ControlBoardWrapper| /icubSim/torso  initting YARP initialization
[INFO] |yarp.os.Port| Port /icubSim/torso/rpc:i active at tcp://192.168.1.84:10008/
[INFO] |yarp.os.Port| Port /icubSim/torso/command:i active at tcp://192.168.1.84:10009/
[INFO] |yarp.os.Port| Port /icubSim/torso/state:o active at tcp://192.168.1.84:10010/
[INFO] |yarp.os.Port| Port /icubSim/torso/stateExt:o active at tcp://192.168.1.84:10011/
[INFO] |yarp.dev.PolyDriver| Created wrapper <controlboardwrapper2>. See C++ class ControlBoardWrapper for documentation.
[DEBUG] found:  torso_yaw torso_yaw
[DEBUG] found:  torso_pitch torso_pitch
[DEBUG] found:  torso_roll torso_roll
[DEBUG] done
[WARNING] No positionToleranceLinear value found in ini file, default one will be used!
[WARNING] No positionToleranceRevolute value found in ini file, default one will be used!
[WARNING] No max torques value found in ini file, default one will be used!
[DEBUG] max_torques: [   2000.000000	 2000.000000	 2000.000000  ]
[INFO] min_stiffness param found!
[INFO] max_stiffness param found!
[INFO] min_damping param found!
[INFO] max_damping param found!
[DEBUG] min_stiffness: [   0.000000	 0.000000	 0.000000  ]
[DEBUG] max_stiffness: [   1000.000000	 1000.000000	 1000.000000  ]
[DEBUG] min_damping: [   0.000000	 0.000000	 0.000000  ]
[DEBUG] max_damping: [   100.000000	 100.000000	 100.000000  ]
[WARNING] VELOCITY_CONTROL: 'velocityControlImplementationType' param missing.  Using the default value of 'direct_velocity_pid', but the parameter will be compulsory gazebo-yarp-plugins 4.
[DEBUG] Initializing Trajectory Generator with current values
[INFO] |yarp.dev.PolyDriver| Created device <gazebo_controlboard>. See C++ class GazeboYarpControlBoardDriver for documentation.
[Wrn] [Publisher.cc:136] Queue limit reached for topic /gazebo/default/user_camera/pose, deleting message. This warning is printed only once.
```

**Comment by traversaro** [comment]
Indeed, by directly inspecting the iCubGazebo9 and iCubV3 models it is possible to see that those models only contain the upper body links, no sign of the lower bodies.

**Comment by traversaro** [comment]
It seems that the iCubV3 and iCubGazebo9 models are both generated from the upperbody models, see the files used in: https://github.com/robotology/icub-model-generator/blob/master/simmechanics/CMakeLists.txt#L285 . Probably we should switch that to use the all body option, using the files provided in https://github.com/robotology/icub-model-generator/tree/master/simmechanics/data/icub3 ?

**Comment by Nicogene** [comment]
Sorry, I didn't expect that the deployment gave so many troubles, I couldn't check thess issues since I always worked inside `icub-model-generators`, I will work with @traversaro in order to fix the model inside `icub-models`, in the meanwhile here is the model if you need to try it.
Probably you would need to copy by hand the conf files and meshes in the installation folders.

[icub3-all.zip](https://github.com/robotology/icub-models/files/5399674/icub3-all.zip)

**Comment by S-Dafarra** [comment]
I can load the ``iCubV3`` model now :tada: :tada: 
![image](https://user-images.githubusercontent.com/18591940/96445095-a101ef00-120f-11eb-895f-23ef075642c7.png)

It seems that the arms are not controlled though

**Comment by Nicogene** [comment]
> It seems that the arms are not controlled though

Yes because in the asm we have only the shoulders, we thought to create yet another conf file for controlling only the shoulders like `right/left_arm_no_forearm.ini` but in practice it may be not so useful since we are going to add the forearms relatively soon.

It it gives any problem we can add those files ore make the shoulders fixed

**Comment by S-Dafarra** [comment]
> Yes because in the asm we have only the shoulders, we thought to create yet another conf file for controlling only the shoulders like `right/left_arm_no_forearm.ini` but in practice it may be not so useful since we are going to add the forearms relatively soon.

Ah ok, I see. I think we are going to use it extensively starting from Wednesday, do you think that adding the forearms may require more time?

**Comment by Nicogene** [comment]
We decided to exploit the momentum and we added also the arms 💪 

![immagine](https://user-images.githubusercontent.com/19152494/96595050-8ef57e80-12eb-11eb-90d9-ea79c24abecd.png)


The PR is https://github.com/robotology/icub-model-generator/pull/147

The model changed the name from `iCubV3` to `iCubGazeboV3` since we had to fake the arms inertias for this issue https://github.com/robotology/icub-models/issues/53.

The model should be soon available on `icub-models@devel`.

cc @maggia80 @pattacini @S-Dafarra @DanielePucci @salvi-mattia @traversaro

**Comment by pattacini** [comment]
For the record, @GiulioRomualdi and @S-Dafarra managed to get iCub 3 (preliminarily) walking w/ the latest URDF ✨ 

![icub3-walking](https://user-images.githubusercontent.com/3738070/96963591-2e438d00-1509-11eb-928b-2f5483dd9ca2.gif)

**Comment by GiulioRomualdi** [comment]
Here I drop some issues related to the generation of the model:
https://github.com/robotology/icub-models/issues/60
https://github.com/robotology/icub-models/issues/61
https://github.com/robotology/icub-models/issues/57

**Comment by Nicogene** [comment]
This PR should(https://github.com/robotology/icub-model-generator/pull/163) reduce the weight of the feet, wrongly exported by the magic button

**Comment by traversaro** [comment]
I think the improvements on the iCub3 model will be constant for the next few years, so what do you think if we close this issue and open separate ones to track the particular improvements?

**Comment by S-Dafarra** [comment]
> I think the improvements on the iCub3 model will be constant for the next few years, so what do you think if we close this issue and open separate ones to track the particular improvements?

Ok for me

**Comment by traversaro** [comment]
As agreed, let's close the issue, to check all the issues related to iCub3 it is possible to check out the `icub3` label: https://github.com/robotology/icub-models/labels/icub3 .
